src/backend/distributed/operations/rebalancer_placement_isolation.c

```diff
@@ -0,0 +1,474 @@
+/*-------------------------------------------------------------------------
+ *
+ * rebalancer_placement_isolation.c
+ *	  Routines to determine which worker node should be used to separate
+ *	  a colocated set of shard placements that need separate nodes.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "nodes/pg_list.h"
+#include "utils/hsearch.h"
+#include "utils/lsyscache.h"
+
+#include "distributed/colocation_utils.h"
+#include "distributed/hash_helpers.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/metadata_utility.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/rebalancer_placement_isolation.h"
+#include "distributed/shard_rebalancer.h"
+
+
+struct RebalancerPlacementIsolationContext
+{
+	HTAB *nodePlacementGroupHash;
+};
```

Comment:
This struct is defined to prevent callers from manipulating htab somehow. That is why the header file contains the typedef and .c contains struct definition, in the end to to provide an opaque def.